### ISSUE TITLE
Database connection pooling for MySQL & PostgreSQL external DBs

### ISSUE THREAD

**Issue by dnoland1** [rancher/k3s#1459]
**Is your feature request related to a problem? Please describe.**
Yes, currently seeing "Error rpc error: code = Unknown desc = Error 1040: Too many connections" in the UI:
![image](https://user-images.githubusercontent.com/468462/75395374-21ea3000-58a7-11ea-83f2-d86b794abb79.png)

Appears this is due to hitting the maximum number of database connections. Currently using an AWS RDS Aurora database (type db.t3.small) with `max_connections = 100`. Have 2 k3s servers, 0 agents in the cluster. Error frequently occurs when navigating to the local -> System project. Running Rancher v2.3.5.

**Describe the solution you'd like**
Would like to have a connection limit or database connection pooling implemented in k3s so that each cluster will have a predictable maximum number of DB connections open on the database server. May run multiple k3s clusters against a single RDS cluster and need to know the expected number of connections. If it makes sense, this max or pool size would be a configurable parameter with a reasonable limit.

**Describe alternatives you've considered**
Arbitrarily raising database max_connections on the server to avoid the error.

**Comment by ibuildthecloud** [comment]
My suspicion is this is really a issue with rancher. Is rancher running on the DB backed k3s?

We can put in a conn limit but that does just mean things will be slower. It's probably better to fix Rancher. 2.3 is known to be overly chatty and has already been addressed in the dev versions of Rancher.

**Comment by dnoland1** [comment]
Yes, Rancher is running on the k3s being backed by RDS. Is it worth putting in an issue in rancher/rancher to have the overly chatty fixes backported to v2.3? Will likely move to v2.4 once it GAs.

**Comment by deltasquare4** [comment]
I am having the same problem with RDS PostgreSQL backed k3s cluster. Removing Rancher did not help.

Also, this probably falls under [Kine](https://github.com/rancher/kine) implementation now?

**Comment by MarkDong614** [comment]
I am having the same problem with mysql backed k3s cluster.

**Comment by izkuru** [comment]
I am also having this same problem, in a k3s cluster backed by mysql (`v1.17.2+k3s1`). From my experience in my cluster, it gets into a state when 2 of 3 masters are behaving fine, but the 3rd master starts going haywire, attempting to open thousands of connections (which looks to me that there is no upper bound for the connection pool).

The `MySQL/MariaDB` section on https://rancher.com/docs/k3s/latest/en/installation/datastore/ links to the docs for the `go-sql-driver` for mysql for advanced options: https://github.com/go-sql-driver/mysql#dsn-data-source-name

This in turn points to the the golang `database/sql` documentation for the connection pool settings, specifically, these settings:
https://golang.org/pkg/database/sql/#DB.SetConnMaxLifetime
https://golang.org/pkg/database/sql/#DB.SetMaxIdleConns
https://golang.org/pkg/database/sql/#DB.SetMaxOpenConns

It looks like the default behavior for `database/sql` is that `MaxOpenConns` is set to 0 (unlimited). 

> Also, this probably falls under [Kine](https://github.com/rancher/kine) implementation now?

That was my assumption as well, as it likes like kine is the wrapper no matter what your cluster datastore backend is. And looking over the source code for kine, it doesn't look like there is anything that currently sets the parameters for the `database/sql` library.  So, essentially, the k3s service has no upper limit on how many connections it will try and open with the backend, and can lead to the database being overloaded if circumstances are just right. This lines up with the behavior I've seen my cluster exhibiting. 

Not only would kine need to be adjusted to set these parameters, we would also need some way to influence the values from k3s as well.

**Comment by ahazelwood** [comment]
I am seeing the same issues with using mysql.  In my case I have just 2 masters, and the number of connections opened result in mysql logging: 

`Aborted connection 2888824 to db: 'k3s_search' user: 'k8s' host: 'leader02' (Got an error reading communication packets)
`

MySQL version: `mysql  Ver 14.14 Distrib 5.7.29, for Linux (x86_64)`
k3s Cluster: `v1.17.4+k3s1`

**Comment by dbecher-ito** [comment]
I created a PR for kine here https://github.com/rancher/kine/pull/37 to tackle the issue.

When testing rancher installed on a k3s AWS instance I also stumbled upon this error messages - I used a fairly small AWS db.t2.micro instance for SQL which could not always handle displaying e.g. the workloads in the System project.

The PR is a proposal of how to make connection pooling actually configurable.

Locally I also re-compiled k3s with the proposed changes (had to add configurable cli flags in [https://github.com/rancher/k3s/blob/master/pkg/cli/cmds/server.go](https://github.com/rancher/k3s/blob/master/pkg/cli/cmds/server.go) to test this locally) and it worked! I limited the maximum connections and the error vanished 🎉 

For this repo, when the PR in the kine project would be accepted, it should also expose the mentioned cli flags on the `k3s server`, I imagine something like this:

| Flag                                  | Default        | Description                                                                                                                     |
|:--------------------------------------|:---------------|:--------------------------------------------------------------------------------------------------------------------------------|
| `--datastore-max-idle-connections`    | `2`            | Maximum number of idle connections used by datastore. If num <= 0, then no connections are retained                             |
| `--datastore-max-open-connections`    | `0`(unlimited) | Maximum number of idle connections used by datastore. If num <= 0, then there is no limit                                       |
| `--datastore-connection-max-lifetime` | `1s`           | Maximum duration a connection is held alive. Desfined as a parsable string, e.g., 1s, 2500ms, and 1h30m are all accepted values |

Maybe even make them configurable via environment variables?

I would create a PR for this repo as well with my changes for the server cmd, but since its depending on the kine project, I think the PR should be merged there first before I would create a PR for this project.

Cheers!

**Comment by CarpathianUA** [comment]
Experiencing the same. k3s 1.18 with AWS RDS MySQL as a datastore.

**Comment by spicysomtam** [comment]
We are seeing this as well. k3s. mysql backend using a aws t2.micro rds.

**Comment by fabinou** [comment]
Hello

I have the same issue using MySQL on another server (Debian 10 MariaDB for the two server).

 Is there a workaround for this, because right now, i just can't use rancher. It's up for a few second, then down, and so on ...

And of course i have a lot of :

`2020-06-15 14:32:47 21311 [Warning] Aborted connection 21311 to db: 'k3sbase' user: 'k3s-base' host: 'xxx.xxx.xxx.xxx' (Got an error reading communication packets)
2020-06-15 14:32:49 21312 [Warning] Aborted connection 21312 to db: 'k3sbase' user: 'k3s-base' host: 'xxx.xxx.xxx.xxx' (Got an error reading communication packets)
2020-06-15 14:32:57 21256 [Warning] Aborted connection 21256 to db: 'k3sbase' user: 'k3s-base' host: 'xxx.xxx.xxx.xxx' (Got an error reading communication packets)
2020-06-15 14:32:59 21310 [Warning] Aborted connection 21310 to db: 'k3sbase' user: 'k3s-base' host: 'xxx.xxx.xxx.xxx' (Got an error reading communication packets)
2020-06-15 14:33:01 21316 [Warning] Aborted connection 21316 to db: 'k3sbase' user: 'k3s-base' host: 'xxx.xxx.xxx.xxx' (Got an error reading communication packets)
`

**Comment by brandond** [comment]
Possibly related: https://github.com/rancher/k3s/issues/1912#issuecomment-645844326

I would like to see connection pooling made available though, and perhaps enabled by default if it turns out to improve performance.

**Comment by brandond** [comment]
@fabinou do you have error logs from the client side? I saw the same messages on the DB server when k3s crashed due to queries taking way too long to complete. If the lease updates don't complete in time the controller manager goroutine will exit its loop causing k3s to terminate, and you get messages like that on the DB side as it closes all the open connections uncleanly.

**Comment by dbecher-ito** [comment]
@fabinou 
> Hello
> 
> I have the same issue using MySQL on another server (Debian 10 MariaDB for the two server).

Your maximum number of connections seems high enough, so I assume your posted error messages have another cause and is not related to this issue.

**Comment by zhangalex** [comment]
I think it is caused by 'SELECT COUNT(*) FROM kine;' when k3s trying to start. The query stuck the whole MySQL server, especially when your DB is a micro/small instance.

**Comment by ahazelwood** [comment]
I don't believe the size has anything to do with it.  We were using it a full-blown VM instance on-premises with 6 cores and 16GB of RAM.

**Comment by davidnuzik** [comment]
Based on testing, the connection count seems to increase when there is a slow disk. Further, we have observed that as we fall behind (slow queries) this will "snowball" and get worse if the issue continues.
Additionally, if we exceed the max_connections we seem to fail hard.

We could add a tunable for connection pooling, which could help mitigate the issue. @brandond any other remarks?

**Comment by fredleger** [comment]
hit the same on k3s version v1.19.10+k3s1 (72e8196c) (fresh install) with google cloud sql instance postgresql 13 (also fresh) / db-f1-micro
Bellow are the repartition of the error "pq: remaining connection slots are reserved for non-replication superuser connections" when k3s start. Rancher isn't able to start cleanly.

![image](https://user-images.githubusercontent.com/2778741/116228696-ed496100-a755-11eb-8bbf-b3894488299a.png)

According to this doc https://cloud.google.com/sql/docs/postgres/quotas the max conn is 25

**Comment by fredleger** [comment]
it seems that there is one request that impact a lot the connection usage time : 

```SQL
SELECT
  (
  SELECT
    MAX(rkv.id) AS id
  FROM
    kine AS rkv),
  (
  SELECT
    MAX(crkv.prev_revision) AS prev_revision
  FROM
    kine AS crkv
  WHERE
    crkv.name = $3),
  kv.id AS theid,
  kv.name,
  kv.created,
  kv.deleted,
  kv.create_revision,
  kv.prev_revision,
  kv.lease,
  kv.value,
  kv.old_value
FROM
  kine AS kv
JOIN (
  SELECT
    MAX(mkv.id) AS id
  FROM
    kine AS mkv
  WHERE
    mkv.name LIKE $1
  GROUP BY
    mkv.name) maxkv
ON
  maxkv.id = kv.id
WHERE
  (kv.deleted = $4
    OR $2)
ORDER BY
  kv.id ASC
LIMIT
  $5
```

**Comment by fredleger** [comment]
It's fixed if i use a larger db instance.
So i think it worth a note in the docs to mention required db if not present (i didn't notice it) and a switch to limit connections would be nice.

**Comment by david-becher** [comment]
I created a pull request to kine a while ago which addresses exactly that: database connection pooling. See especially this struct https://github.com/k3s-io/kine/blob/master/pkg/drivers/generic/generic.go#L70. The only thing left is another pull request on this repository, to create command line flags/environment vars for k3s, which control parameters of the aforementioned struct.

**Comment by fredleger** [comment]
hi @david-becher 

Yes we indeed saw it and event tried the flags, not really getting if this was still a proposal or a merged change in k3s.
We came to the understanding that code was merged and the only thing missing were the flags as you confirmed it.

So let's keep this open and active to have the flags available very soon then ;-)

For the records we have to go for a postgress instance to get around this point just because the flags are not supported yet. This is a bit sad since the guys at the company don't know much PG compared to MySQL/Mariadb.

**Comment by dereknola** [comment]
After compiling K3s with the exposed database arguments, testing has determine that modifying the max connections and idle connections has not produced a meaningful improvement in DB performance. A slow DB with still cause large numbers of queries and connections to exist due to slow response time from the DB on those queries. As such, we will not be exposing these additional toggles to K3s as it simply obfuscates a larger issue.

**Comment by vasicvuk** [comment]
@dereknola I configured Postgres based on hardware requirements specified in the documentation and this problem occurs very fast, This means that sizing in docs is wrong. Currently, load of Postgres is 50 on the 2 vCPU and 8gb machine for only 24 nodes (2 masters)
